keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1709 @@
+
+### Unions
+
+Unions are informally defined but common in built-in types.
+
+We will convert most union validation to CEL expressions. For example:
+
+```go
+// staging/src/k8s.io/api/apps/v1/types.go
+
+type DeploymentSpec struct {
+  // ...
+  //
+  // +minimum=0
+  Replicas string `json...`
+  
+  // ...
+  // 
+  // +validationRule="!(self.type == 'Recreate') || !has(self.rollingUpdate)",message="may not be specified when strategy `type` is Recreate",reason=Forbidden,field="rollingUpdate"
+  // +validationRule="!(self.type == 'RollingUpdate') || has(self.rollingUpdate)",message="this should be defaulted and never be nil",reason=Required,field="rollingUpdate"
+  Strategy DeploymentStrategy `json:...`
+  
+  // ...
+}
+```
+
+For backward compatibility with existing validations, `reason` and `field` will
+be added to make it possible to declare a validation rule in CEL that matches
+not just the logic of the hand written validation it replaces, but also the
+exact field path and reason type.
+
+### CEL Validation
+
+CEL will play a major role in declarative validation by offering a way to
+declare validation rules for use cases that are too complex to be declared using
+JSON Schema value validations. Using CEL to validate APIs has been successfully
+demonstrated by CRD Validation Rules, which are on track for GA in 1.28.
+
+CEL rules will typically be placed on struct or field declarations and access
+multiple fields nested below the level of the type or field where the CEL
+expression is declared, e.g.:
+
+```go
+  
+  //+validationRule="!self.widgetType == 'Component' || !['badname2', 'badname2'].exists(notAllowed, self.componentName.contains(notAllowed))",reason=Forbidden
+  type FizzBuzzSpec struct {
+    Type WidgetType `json...`
+    ComponentName string `json...`
+  }
+```
+
+Note that the `+validationRule` IDL tags supports multiple optional fields after the expression string. The supported fields are:
+
+- `message`: A vailidation failure message string.
+- `messageExpression`: A CEL expression that evalutes to a validation failure message string. Takes precedence over message.
+- `reason`: Must be one of [Required, Forbidden, Invalid, RequestEntityTooLarge].
+- `field`: A JSONPath to the sub-field that a validation failure should be attributed to.
+
+We will need to extend our CEL libraries to make it possible to migrate all the
+validation rules that exist in the Kubernetes API today.
+
+- `isFormat() <bool>` and `validateFormat() <list<string>>` will be added to allow formats to be checked in CEL
+  expression and for format violations to be reported using `messageExpression: "self.validateFormat('ipv6')"`
+- IP and CIDR library:
+  - `isIP(string) bool`
+  - `ip(string) IP`
+  - `IP.is4() bool`
+  - `IP.is6() bool`
+  - `IP.isLoopback() bool`
+  - `isCIDR(string) bool`
+  - `cidr(string) CIDR`
+  - `CIDR.overlaps(CIDR) bool`
+  - `CIDR.containsIP(IP) bool`
```

Comment:
+ @thockin @robscott for ideas on what specifically we should support for IPs and CIDRs.  Also, are there other network concepts we need better direct support for?

Reply:
I think we need to audit a bunch of hand-written code and brainstorm the least-painful CEL equivalents and for things that are just TOO painful, derive new helper functions.  See my comment above.  As a non-CEL expert, I don't know all of what's viable.

Reply:
@thockin CEL team would be happy to accept a package for working with CIDR masks. We try to remain compatible with ZetaSQL, so if there's a [subset of functions](https://github.com/google/zetasql/blob/master/docs/net_functions.md) from there which would be desired here, we can iterate on a CEL-idiomatic way of exposing them. If you need more than what's listed, we can also support that too, but would like to align where possible.

In general, even for other packages, I can consult on what is general enough to include in CEL and what should probably be K8s-specific.